index-scheduler/src/batch.rs

```diff
@@ -689,7 +696,67 @@ impl IndexScheduler {
 
                 Ok(tasks)
             }
+            Batch::IndexSwap { mut task } => {
+                let mut wtxn = self.env.write_txn()?;
+                let swaps = if let KindWithContent::IndexSwap { swaps } = &task.kind {
+                    swaps
+                } else {
+                    unreachable!()
+                };
+                for (lhs, rhs) in swaps {
+                    self.apply_index_swap(&mut wtxn, task.uid, lhs, rhs)?;
+                }
+                wtxn.commit()?;
+                task.status = Status::Succeeded;
+                Ok(vec![task])
+            }
+        }
+    }
+
+    /// Swap the index `lhs` with the index `rhs`.
+    fn apply_index_swap(&self, wtxn: &mut RwTxn, task_id: u32, lhs: &str, rhs: &str) -> Result<()> {
+        // 1. Verify that both lhs and rhs are existing indexes
+        let index_lhs_exists = self.index_mapper.index_exists(&wtxn, lhs)?;
+        if !index_lhs_exists {
+            return Err(Error::IndexNotFound(lhs.to_owned()));
         }
+        let index_rhs_exists = self.index_mapper.index_exists(&wtxn, rhs)?;
+        if !index_rhs_exists {
+            return Err(Error::IndexNotFound(rhs.to_owned()));
+        }
+
+        // 2. Get the task set for index = name.
+        let mut index_lhs_task_ids =
+            self.get_task_ids(&Query::default().with_index(lhs.to_owned()))?;
+        index_lhs_task_ids.remove_range(task_id..);
+        let mut index_rhs_task_ids =
+            self.get_task_ids(&Query::default().with_index(rhs.to_owned()))?;
+        index_rhs_task_ids.remove_range(task_id..);
```

Comment:
Why not directly use the `Query` system to only retrieve the non-enqueued tasks?